src/api.rs

```diff
@@ -928,10 +929,26 @@ impl<T: Pixel> ContextInner<T> {
       .clone()
   }
 
+  /// Indicates whether more frames need to be read into the frame queue
+  /// in order for lookahead to be full.
   pub(crate) fn needs_more_lookahead(&self) -> bool {
-    self.needs_more_frames(self.frame_count)
-      && self.frames_processed + LOOKAHEAD_FRAMES
-        > self.frame_q.keys().last().cloned().unwrap_or(0)
+    let lookahead_end = self.frame_q.keys().last().cloned().unwrap_or(0);
+    let last_frame_processed = self.frame_invariants.values()
+      .filter(|fi| fi.show_frame)
+      .filter(|fi| !fi.invalid)
+      .map(|fi| fi.input_frameno)
+      .max()
+      .unwrap_or(0);
+    self.needs_more_frames(lookahead_end)
+      && lookahead_end < last_frame_processed + LOOKAHEAD_FRAMES
+  }
+
+  /// Indicates whether the lookahead queue is ready for frames to be
+  /// analyzed for scene changes.
+  pub(crate) fn lookahead_ready(&self) -> bool {
+    let lookahead_end = self.frame_q.keys().last().cloned().unwrap_or(0);
+    self.next_lookahead_frame + LOOKAHEAD_FRAMES < lookahead_end ||
```

Comment:
Unrelated: Do we intend LOOKAHEAD_FRAMES to always be hardcoded?

Reply:
We could make it a variable and check what breaks if we set it to 2.

Reply:
In the future it will probably be a CLI option. We have access to `self.config` from here, I believe, so that should be fine.